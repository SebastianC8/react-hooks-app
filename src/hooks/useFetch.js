import { useEffect, useRef, useState } from "react"

export const useFetch = (url) => {

    const [state, setState] = useState({ data: null, loading: true, error: null });
    const isMounted = useRef(true);

    useEffect(() => {

        /**
         * Cuando se desmonta el componente
         */
        return () => {
            isMounted.current = false;
        }

    }, []);

    useEffect(() => {

        /**
         * Cuando se ejecuta el componente
         */
        setState({ data: null, loading: true, error: null });

        fetch(url)
        .then((response) => response.json())
        .then((data) => {

            if (isMounted.current) {
                setState({ data: data, loading: false, error: null });
            } else {
                console.log('No se llam√≥');
            }

        });
        
    }, [url]);

    return state;

}