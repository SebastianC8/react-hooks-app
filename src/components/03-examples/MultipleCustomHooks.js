import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';
import '../02-useEffect/effects.css';

export const MultipleCustomHooks = () => {

    const { state, increment } = useCounter(1);
    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${state}`);

    /**
     * Si existe la data, asigna los valores de data
     */
    const { author, quote } = !!data && data[0];

    return (
        <div>
            <h1>Breaking bad quotes - useFetch</h1>
            <hr/>

            {
                loading
                ?
                (
                    <div className="alert alert-info text-center">Loading...</div>
                )
                :
                (
                    <blockquote className="blockquote text-right">
                        <p className="mb-0">{quote}</p>
                        <footer className="blockquote-footer mt-0">{author}</footer>
                    </blockquote>
                )
            }

            <button className='btn btn-primary' onClick = { increment }>Siguiente frase</button>

        </div>
    )
}
