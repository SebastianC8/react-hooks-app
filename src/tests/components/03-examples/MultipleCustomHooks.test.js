import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
jest.mock('../../../hooks/useFetch');

describe('Prueba MultipleCustomHook', () => {
   
    test('should show ok', () => {
        
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks/>);
        expect(wrapper).toMatchSnapshot();

    });

    test('Debe mostrar la información', () => {
        
        useFetch.mockReturnValue({
            data: [{
                author: 'sebastián c.',
                quote: 'Hola mundo'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks/>);
        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text().trim()).toBe('Hola mundo');
        expect(wrapper.find('footer').text().trim()).toBe('sebastián c.');
    });

});
