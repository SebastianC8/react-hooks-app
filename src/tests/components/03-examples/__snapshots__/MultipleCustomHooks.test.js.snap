// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba MultipleCustomHook should show ok 1`] = `
<div>
  <h1>
    Breaking bad quotes - useFetch
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</div>
`;
